drv/grapefruit-seq-server/src/main.rs

```diff
@@ -0,0 +1,266 @@
+                // we should never be notified --- and if one never wakes up,
+                // the difference between sleeping and dying seems kind of
+                // irrelevant. But, `rustc` doesn't realize that this should
+                // never return, we'll stick it in a `loop` anyway so the main
+                // function can return `!`
+                sys_recv_notification(0);
+            }
+        }
+    }
+}
+
+struct ServerImpl<S: SpiServer> {
+    sys: sys_api::Sys,
+    seq: SpiDevice<S>,
+}
+
+const FAULT_PIN_L: sys_api::PinSet = sys_api::Port::A.pin(15);
+
+const FPGA_PROGRAM_L: sys_api::PinSet = sys_api::Port::B.pin(5);
+const FPGA_INIT_L: sys_api::PinSet = sys_api::Port::B.pin(6);
+const FPGA_CONFIG_DONE: sys_api::PinSet = sys_api::Port::B.pin(4);
+
+impl<S: SpiServer + Clone> ServerImpl<S> {
+    fn init(
+        sys: &sys_api::Sys,
+        spi: S,
+        aux: drv_auxflash_api::AuxFlash,
+    ) -> Result<Self, SeqError> {
+        // Ensure the SP fault pin is configured as an open-drain output, and pull
+        // it low to make the sequencer restart externally visible.
+        sys.gpio_configure_output(
+            FAULT_PIN_L,
+            sys_api::OutputType::OpenDrain,
+            sys_api::Speed::Low,
+            sys_api::Pull::None,
+        );
+        sys.gpio_reset(FAULT_PIN_L);
+
+        // To allow for the possibility that we are restarting, rather than
+        // starting, we take care during early sequencing to _not turn anything
+        // off,_ only on. This means if it was _already_ on, the outputs should
+        // not glitch.
+
+        // To program the FPGA, we're using "slave serial" mode.
+        //
+        // See "7 Series FPGAs Configuration", UG470 (v1.17) for details,
+        // as well as "Using a Microprocessor to Configure Xilinx 7 Series FPGAs
+        // via Slave Serial or Slave SelectMAP Mode Application Note" (XAPP583)
+
+        // Configure the PROGRAM_B line to the FPGA
+        sys.gpio_set(FPGA_PROGRAM_L);
+        sys.gpio_configure_output(
+            FPGA_PROGRAM_L,
+            sys_api::OutputType::OpenDrain,
+            sys_api::Speed::Low,
+            sys_api::Pull::None,
+        );
+
+        // Pulse PROGRAM_B low for 1 ms to reset the bitstream
+        // (T_PROGRAM is 250 ns, so this is fine)
+        // https://docs.amd.com/r/en-US/ds189-spartan-7-data-sheet/XADC-Specifications
+        sys.gpio_reset(FPGA_PROGRAM_L);
+        hl::sleep_for(1);
+        sys.gpio_set(FPGA_PROGRAM_L);
+
+        // Wait for INIT_B to rise
+        loop {
+            let init = sys.gpio_read(FPGA_INIT_L) != 0;
+            ringbuf_entry!(Trace::Ice40Init(init));
+            if init {
+                break;
+            }
+
+            // Do _not_ burn CPU constantly polling, it's rude. We could also
+            // set up pin-change interrupts but we only do this once per power
+            // on, so it seems like a lot of work.
```

Comment:
This is decidedly *not* a big deal, but note that the code in Gimlet that also has the comment about not using pin-change interrupts because it's "a lot of work" long predates our implementation of EXTI support in `stm32xx-sys`, so I think it would have been a lot more work when that was written.

On the other hand, I don't think we currently have an interface for a task to *release* a claimed EXTI notification, so using EXTI for this would permanently occupy our ability to receive interrupts for pin changes on port B, so...we probably shouldn't do that. It might be a good idea to have a way for a task to release its configured EXTI interrupt, so that we can use EXTI for one shot stuff like this without tying up that port forever...